SentinelOne/sentinelone_module/singularity/connectors.py

```diff
@@ -0,0 +1,136 @@
+            api_token=self.module.configuration.api_token,
+        )
+
+    def stop(self, *args: Any, **kwargs: Optional[Any]) -> None:
+        """
+        Stop the connector.
+
+        Temporary redefine the method to avoid known SDK issues.
+        """
+        super(Connector, self).stop(*args, **kwargs)
+
+    async def single_run(self) -> int:
+        result = 0
+
+        last_event_date = self.last_event_date.offset
+        while self.running:
+            cursor: str | None = self.last_checkpoint.offset
+            if cursor == "":  # TODO: Fix this in SDK in cursor handling.
+                cursor = None
+
+            start_time = last_event_date.timestamp() if not cursor else None
+            data = await self.client.list_alerts(
+                product_name=self.product_name,
+                after=cursor,
+                start_time=start_time,
+            )
+
+            pushed_events = await self.push_data_to_intakes(
+                [orjson.dumps(alert).decode("utf-8") for alert in data.alerts]
+            )
+
+            result += len(pushed_events)
+
+            for alert in data.alerts:
+                alert_detected_at = isoparse(alert["detectedAt"])
+                if alert_detected_at > last_event_date:
+                    last_event_date = alert_detected_at
+
+            self.last_checkpoint.offset = data.end_cursor
+
+            if not data.has_next_page:
+                break
+
+        self.last_checkpoint.offset = ""
+        self.last_event_date.offset = last_event_date
+
+        return result
+
+    async def async_run(self) -> None:  # pragma: no cover
+        while self.running:
+            try:
+                processing_start = time.time()
+                result = await self.single_run()
+                last_event_date = self.last_event_date.offset
+                processing_end = time.time()
+
+                EVENTS_LAG.labels(intake_key=self.configuration.intake_key).set(
+                    processing_end - last_event_date.timestamp()
+                )
+
+                log_message = "No records to forward"
+                if result > 0:
+                    log_message = "Pushed {0} records".format(result)
+
+                self.log(message=log_message, level="info")
+                OUTCOMING_EVENTS.labels(intake_key=self.configuration.intake_key).inc(result)
+
+                processing_time = processing_end - processing_start
+                FORWARD_EVENTS_DURATION.labels(intake_key=self.configuration.intake_key).observe(processing_time)
+                logger.info(
+                    "Processing took {processing_time} seconds",
+                    processing_time=processing_time,
+                )
+
+                if result == 0:
+                    await asyncio.sleep(self.configuration.frequency)
+
+            except Exception as error:
+                self.log_exception(error)
+                await asyncio.sleep(self.configuration.frequency)
```

Comment:
Can you implement somehting like [this](https://github.com/SEKOIA-IO/automation-library/blob/main/Mimecast/mimecast_modules/connector_mimecast_siem.py#L157-L178) to handle authentication and permissions errors?
The purpose is to log and stop the connector if the credentials are invalid or the permissions denied.